### PR TITLE
Make block.Stream() return a copy of data, even in the case where it performs a merge internally.

### PR DESCRIPTION
**Pull request by richardartoul**

### CODE REVIEW

File: src/dbnode/storage/block/block.go [hunk 1]
```diff
@@ -230,7 +230,11 @@ func (b *dbBlock) Stream(blocker context.Context) (xio.BlockReader, error) {
 	lockUpgraded = true
 	b.Lock()
 
-	// NB: need to re-check as we just upgraded the lock.
+	// Need to re-check everything since we upgraded the lock.
+	if b.closed {
```
Comment (prateek) [inline src/dbnode/storage/block/block.go]: can you add a test case for this too

File: src/dbnode/storage/block/block.go [hunk 2]
```diff
@@ -354,10 +363,10 @@ func (b *dbBlock) streamWithRLock(ctx context.Context) (xio.BlockReader, error)
 // TODO(rartoul): The existing ctx is still holding a reference to the old segment so that will hang around
 // and waste memory until the block is closed. We could improve this by swapping out the underlying ctx with
 // a new one, allowing us to close the old one and release the old segment, freeing memory.
-func (b *dbBlock) forceMergeWithLock(ctx context.Context, stream xio.SegmentReader) (xio.BlockReader, error) {
+func (b *dbBlock) forceMergeWithLock(ctx context.Context, stream xio.SegmentReader) error {
```
Comment (prateek) [inline src/dbnode/storage/block/block.go]: you can delete the comment above this yea?

File: src/dbnode/storage/block/block.go (hunk 1 above)
Comment (richardartoul) [reply src/dbnode/storage/block/block.go]: Should I just do a concurrency test? I can't think of any other way...

File: src/dbnode/storage/block/block.go (hunk 2 above)
Comment (richardartoul) [reply src/dbnode/storage/block/block.go]: woosp yeah meant to do that

File: src/dbnode/storage/block/block.go (hunk 1 above)
Comment (prateek) [reply src/dbnode/storage/block/block.go]: yeah it'll have to be. can do the same thing we do in the shard where we have a `testXYZ` that kicks off two go routines calling the different methods, and run `testXYZ` a bunch of times

**Comment by prateek** [comment]
```
=== RUN   TestDatabaseBlockMergeRace
==================
WARNING: DATA RACE
Write at 0x00c4216c2840 by goroutine 46:
  github.com/m3db/m3db/src/dbnode/storage/block.TestDatabaseBlockMergeRace.func1.1()
      /home/runner/workspace/src/github.com/m3db/m3db/src/dbnode/storage/block/block_test.go:265 +0x29e

Previous write at 0x00c4216c2840 by goroutine 41:
  github.com/m3db/m3db/src/dbnode/storage/block.TestDatabaseBlockMergeRace.func1.1()
      /home/runner/workspace/src/github.com/m3db/m3db/src/dbnode/storage/block/block_test.go:265 +0x29e

Goroutine 46 (running) created at:
  github.com/m3db/m3db/src/dbnode/storage/block.TestDatabaseBlockMergeRace.func1()
      /home/runner/workspace/src/github.com/m3db/m3db/src/dbnode/storage/block/block_test.go:280 +0x935
  github.com/m3db/m3db/src/dbnode/storage/block.TestDatabaseBlockMergeRace()
      /home/runner/workspace/src/github.com/m3db/m3db/src/dbnode/storage/block/block_test.go:284 +0x1ae
  testing.tRunner()
      /home/runner/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:777 +0x16d

Goroutine 41 (finished) created at:
  github.com/m3db/m3db/src/dbnode/storage/block.TestDatabaseBlockMergeRace.func1()
      /home/runner/workspace/src/github.com/m3db/m3db/src/dbnode/storage/block/block_test.go:280 +0x935
  github.com/m3db/m3db/src/dbnode/storage/block.TestDatabaseBlockMergeRace()
      /home/runner/workspace/src/github.com/m3db/m3db/src/dbnode/storage/block/block_test.go:284 +0x1ae
  testing.tRunner()
      /home/runner/.gimme/versions/go1.10.3.linux.amd64/src/testing/testing.go:777 +0x16d
==================
--- FAIL: TestDatabaseBlockMergeRace (98.97s)
	testing.go:730: race detected during execution of test
```

race in the test code ^^

**Pull request merged**